services/mqtt/service.go

```diff
@@ -0,0 +1,151 @@
+package mqtt
+
+import (
+	"fmt"
+	"log"
+	"sync/atomic"
+	"time"
+
+	pahomqtt "github.com/eclipse/paho.mqtt.golang"
+	"github.com/influxdata/kapacitor/alert"
+)
+
+type QOSLevel byte
+
+const (
+	// best effort delivery. "fire and forget"
+	ATMOSTONCE QOSLevel = iota
+	// guarantees delivery to at least one receiver. May deliver multiple times.
+	ATLEASTONCE
+	// guarantees delivery only once. Safest and slowest.
+	EXACTLYONCE
+)
+
+// DEFAULT_QUIESCE_TIMEOUT is the duration the client will wait for outstanding
+// messages to be published before forcing a disconnection
+const DEFAULT_QUIESCE_TIMEOUT time.Duration = 250 * time.Millisecond
+
+type Service struct {
+	Logger *log.Logger
+
+	configValue atomic.Value
+	client      pahomqtt.Client
+	token       pahomqtt.Token
+}
+
+func NewService(c Config, l *log.Logger) *Service {
+	s := &Service{
+		Logger: l,
+	}
+	s.configValue.Store(c)
+	return s
+}
+
+func (s *Service) config() Config {
+	return s.configValue.Load().(Config)
+}
+
+// TODO(timraymond): improve logging here and in Close
+func (s *Service) Open() error {
+	s.Logger.Println("I! Starting MQTT service")
+
+	c := s.config()
+	opts := pahomqtt.NewClientOptions()
+	opts.AddBroker(c.Broker())
+	opts.SetClientID(c.ClientID) // TODO(timraymond): should we provide a random one?
+	opts.SetUsername(c.Username)
+	opts.SetPassword(c.Password)
+	opts.SetCleanSession(false) // wtf is this? Why does it default to false?
+
+	s.client = pahomqtt.NewClient(opts)
+	s.token = s.client.Connect()
+
+	s.token.Wait()
+
+	if err := s.token.Error(); err != nil {
+		s.Logger.Println("E! Error connecting to MQTT broker at", c.Broker(), "err:", err) //TODO(timraymond): put a legit error in
+		return err
+	}
+	s.Logger.Println("I! Connected to MQTT Broker at", c.Broker())
+	return nil
+
+}
+
+func (s *Service) Close() error {
+	s.client.Disconnect(uint(DEFAULT_QUIESCE_TIMEOUT / time.Millisecond))
+	s.Logger.Println("I! MQTT Client Disconnected")
+	return nil
+}
+
+func (s *Service) Alert(qos QOSLevel, topic, message string) error {
+	s.client.Publish(topic, byte(qos), false, message) // should retained be configureable?
```

Comment:
Ah good catch. I meant to research this some more. Publishers have no guarantee that messages will be received by a subscriber to a topic, since there might not be any clients listening at the time the message was published. In that case, the message is dropped by the broker. Retained (which is the 3rd param in `Publish`) sets a flag on the message that causes the broker to hold onto it and deliver it immediately to clients that subsequently subscribe to that topic. The broker holds onto at most one message per topic, so they're useful in describing state ("light switch is on"). Given this, I think the answer to my question is "yes," and it should probably be configureable from within the TICK script.

This entire series of posts has been invaluable: [MQTT Essentials Part 8: Retained Messages](http://www.hivemq.com/blog/mqtt-essentials-part-8-retained-messages)